### PR TITLE
CA-289145: Close socket if error occurs when connecting

### PR DESCRIPTION
**Pull request by krizex**
Signed-off-by: Yang Qian <yang.qian@citrix.com>

### CODE REVIEW

**Comment by mseri** [comment]
Maybe this should be an opportunity for cleanup and we should replace this io module with the one in xenops (See https://github.com/xapi-project/xenops/pull/16/files#diff-d24db55341bf97301146814e40374e28). @robhoes is there any reason to keep the duplication?

File: lib/io.ml [hunk 1]
```diff
@@ -40,8 +40,12 @@ let write fd buf =
...
   let sockaddr = Unix.ADDR_INET (host, port) in
   let fd = Unix.socket Unix.PF_INET Unix.SOCK_STREAM 0 in
-  Unix.connect fd sockaddr;
-  fd
+  try
+    Unix.connect fd sockaddr;
+    fd
+  with e ->
+    Unix.close fd;
+    raise e
```
Comment (mseri) [inline lib/io.ml]: Now that I think about it, in `xenopsd` we already use `xapi-stdext-pervasives`, so you should probably use `Xapi_stdext_pervasives.Pervasiveext.finally`: `finally (fun () -> Unix.connect fd sockaddr; fd) (fun () -> Unix.close fd)`

Comment (krizex) [reply lib/io.ml]: No, this is not the case that we could apply `finally` here. If no error happens, we should NOT close the `fd`.

Comment (mseri) [reply lib/io.ml]: Oh geez, you are right. I was not thinking straight!

**Pull request merged**